lib/faktory_worker/error_formatter.ex

```diff
@@ -0,0 +1,68 @@
+defmodule FaktoryWorker.ErrorFormatter do
+  @moduledoc false
+
+  defmodule FormattedError do
+    defstruct [:type, :message, :stacktrace]
+  end
+
+  def format_error(error, max_stacktrace_length) do
+    {type, message, stacktrace} = split_error_reason(error)
+
+    message = limit_error_length(message)
+
+    stacktrace =
+      stacktrace
+      |> Enum.take(max_stacktrace_length)
+      |> Enum.map(&Exception.format_stacktrace_entry/1)
+
+    %FormattedError{
+      type: type,
+      message: message,
+      stacktrace: stacktrace
+    }
+  end
+
+  defp split_error_reason({%struct_type{} = struct, stacktrace}) when is_list(stacktrace) do
+    message = Map.get(struct, :message)
+
+    {Atom.to_string(struct_type), message || inspect(struct), stacktrace}
+  end
+
+  defp split_error_reason({%{message: message}, stacktrace})
+       when is_binary(message) and is_list(stacktrace) do
+    {nil, message, stacktrace}
+  end
+
+  defp split_error_reason({{error_type, _} = reason, stacktrace})
+       when is_atom(error_type) and is_list(stacktrace) do
+    {Atom.to_string(error_type), format_reason(reason), stacktrace}
+  end
+
+  defp split_error_reason({reason, stacktrace}) when is_list(stacktrace) do
+    {nil, format_reason(reason), stacktrace}
+  end
+
+  defp split_error_reason(reason) do
+    {nil, format_reason(reason), []}
+  end
+
+  defp format_reason(reason) when is_atom(reason) do
+    Atom.to_string(reason)
+  end
+
+  defp format_reason(reason), do: inspect(reason)
+
+  defp limit_error_length(error) when byte_size(error) > 1000 do
```

Comment:
Is `1000` based on a hard limit from Faktory?

Reply:
Correct, documented in the worker lifecycle wiki page https://github.com/contribsys/faktory/wiki/Worker-Lifecycle#report-result